### ISSUE TITLE
The output size of TransposedConv2DLayer(Deconv2DLayer)

### ISSUE THREAD

**Issue by zzd1992** [Lasagne/Lasagne#746]
My input size is (14,14) and I want the output size to be (28,28) after implementating TransposedConv2DLayer. Here is an example of code:

network = lasagne.layers.InputLayer(shape=(None, 1, 14, 14), input_var=input_var)
network = lasagne.layers.Deconv2DLayer(network, num_filters=1, filter_size=(3, 3), crop='same',
            stride=2,  nonlinearity=lasagne.nonlinearities.rectify)

But the actual output size is (27,27).
Then I read the source code which defines the output size of TransposedConv2DLayer:
(output_length - 1) * stride - 2 * pad + filter_size         (1)

In this example  “output_length” is 14. To get my expected output size, for simplicity, I just change (1) into:
(output_length - 1) * stride - 2 * pad + filter_size + 1   (2)

And it works. But I don't know whether this change is reasonable( I guess the back propagation can't pass through the added dimension). Is there better way to solve this kind of problem?

**Comment by benanne** [comment]
Hi,

This layer should really have an argument for specifying the desired output size, because it is ambiguous. One proposed solution is here: https://github.com/Lasagne/Lasagne/pull/745

Feel free to join the discussion there if you want. But your "hack" should work fine in the meantime :)

**Comment by f0k** [comment]
> ( I guess the back propagation can't pass through the added dimension).

Well, the added dimension can be seen as an extra, one-sided zero padding of the input, after dilation and ordinary symmetric padding. In your case, it will result in an additional output row and column that depends on the outermost input and outermost filter pixels:
```
Input  abcdefghijklmn
Filter ABC

0a b c d e f g h i j k l m n00  # dilated, zero-padded input plus extra zero
ABCABCABCABCABCABCABCABCABCABC  # 9 applications plus extra application
 ABCABCABCABCABCABCABCABCABC    # 9 applications
  ABCABCABCABCABCABCABCABCABC   # 9 applications
```
So there's nothing wrong with it. The output still defaults to 27 because forward convolution of 27 elements with a 3-element filter, pad=same and stride 2 would be enough to give 14 elements:
```
0abcdefghijklmnopqrstuvwxyza0
111 333 555 777 999 111 333
  222 444 666 888 000 222 444
```

> Is there better way to solve this kind of problem?

No, just a better interface, proposed in #711 and hopefully implemented in #745. Closing this!
